packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -895,6 +1080,687 @@ void Add(
+    Ordinal finalPos = 0;
+    Ordinal overPos = 0;
+    while(overPos < rowNum)
+    {
+      //check if the two entries at overPos are the same col
+      if(CcolindOver(rowStart + overPos) == CcolindOver(rowStart + overPos + 1) && (overPos + 1 < rowNum))
+      {
+        //they are the same col, so combine the two entries at finalPos
+        //Then increment overPos an extra time since just processed two entries
+        CcolindOver(rowStart + finalPos) = CcolindOver(rowStart + overPos);
+        CvalsOver(rowStart + finalPos) = CvalsOver(rowStart + overPos) + CvalsOver(rowStart + overPos + 1);
+        overPos++;
+      }
+      else
+      {
+        //they are not the same col, so just copy the one entry at overPos to finalPos
+        CcolindOver(rowStart + finalPos) = CcolindOver(rowStart + overPos);
+        CvalsOver(rowStart + finalPos) = CvalsOver(rowStart + overPos);
+      }
+      overPos++;
+      finalPos++;
+    }
+    Crowcounts(i) = finalPos;
+  }
+  const RowPtrs& CrowptrTemp;
+  RowPtrs& Crowcounts;
+  ColInds& CcolindOver;
+  Values& CvalsOver;
+  //Two scratch arrays, used for sorting (with same size as CcolindOver and CvalsOver)
+  ColInds& CcolindAux;
+  Values& CvalsAux;
+  int numRadixSweeps;
+};
+
+template<typename Ordinal, typename Values, typename RowPtrs, typename ColInds>
+struct CondenseArraysFunctor
+{
+  CondenseArraysFunctor(const RowPtrs& Crowptr_, const RowPtrs& CrowptrTemp_, const ColInds& CcolindOver_, ColInds& Ccolind_, const Values& CvalsOver_, Values& Cvals_) :
+    Crowptr(Crowptr_),
+    CrowptrTemp(CrowptrTemp_),
+    CcolindOver(CcolindOver_),
+    Ccolind(Ccolind_),
+    CvalsOver(CvalsOver_),
+    Cvals(Cvals_)
+  {}
+  //4: Copy values from oversized Cvals/Ccolind to correctly sized
+  KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+  {
+    Ordinal rowNum = Crowptr(i + 1) - Crowptr(i);
+    for(Ordinal j = 0; j < rowNum; j++)
+    {
+      Ccolind(Crowptr(i) + j) = CcolindOver(CrowptrTemp(i) + j);
+      Cvals(Crowptr(i) + j) = CvalsOver(CrowptrTemp(i) + j);
+    }
+  }
+  const RowPtrs& Crowptr;
+  const RowPtrs& CrowptrTemp;
+  const ColInds& CcolindOver;
+  ColInds& Ccolind;
+  const Values& CvalsOver;
+  Values& Cvals;
+};
+}
+
+template<typename SC, typename LO, typename GO, typename NO>
+void AddKernels<SC, LO, GO, NO>::
+Unsorted(
+  const typename AddKernels<SC, LO, GO, NO>::ValuesArray& Avals,
+  const typename AddKernels<SC, LO, GO, NO>::RowPtrsArray& Arowptrs,
+  const typename AddKernels<SC, LO, GO, NO>::ColIndsArray& Acolinds, 
+  const typename AddKernels<SC, LO, GO, NO>::ISC scalarA,
+  const typename AddKernels<SC, LO, GO, NO>::ValuesArray& Bvals,
+  const typename AddKernels<SC, LO, GO, NO>::RowPtrsArray& Browptrs,
+  const typename AddKernels<SC, LO, GO, NO>::ColIndsArray& Bcolinds, 
+  const typename AddKernels<SC, LO, GO, NO>::ISC scalarB,
+  GO numGlobalCols,
+  Teuchos::RCP<typename AddKernels<SC, LO, GO, NO>::ValuesArray>& Cvals,
```

Comment:
See above note on Kokkos::View needing to be stored by value in functors, not by reference.  Also, see above note on RCP vs. Kokkos::View (short answer: just store Kokkos::View; forget about RCP).